Swiching default backend from BDB to SQLite

emerkle826 opened the pull request:
Minor changes to SQLite Xerial implementation to fix some SQL queries. Removed V1 implementation.
Introduced Maven profiles for running stress tests.
Changed WorkingTree implementation for bulk insert to prevent thread deadlock in certain situations (OSM import test).
Added concurrent read/write tests to ObjectDatabase conformance tests to catch concurrency issues as seen with SQLite.

Commit:
Swiching default backend from BDB to SQLite

  Minor changes to SQLite Xerial implementation to fix some SQL queries. Removed V1 implementation.
  Added concurrent read/write tests for ObjectStores to catch concurrency issues as seen with SQLite.

emerkle826 commented:
@groldan This should make SQLite the default and addresses the concurrency issues. I reverted the SQLite driver version back to 3.8.6 to be consistent with the version used in GeoTools. Also added some tests to ensure that concurrency issues are caught in the conformance tests, and not relying totally on the GeoTools DataStore tests.

groldan commented on src/core/src/test/java/org/locationtech/geogig/storage/ObjectStoreConformanceTest.java:
```diff
@@ -77,6 +100,13 @@ public void setUp() throws IOException {
         this.db = createOpen(platform, hints);
         this.db.open();
+
+        dbWriteThreads = Executors.newFixedThreadPool(CONCURRENT_THREAD_COUNT,
```
the thread pools are being created and destroyed even for all the tests that don't use them.
Idea: can we either set them up on the tests that use them, OR, probably better, extract these tests out to a new ObjectStoreConcurrencyTest suite?

groldan commented:
a couple comments:
- please watch out for whitespace only changes, they clutter commit history unnecessarily.
- In WorkingTree.java, we just can't build an ArrayList with all the features being imported. It'll OOM. Hence the in-process conversion. We should get rid of that, can you check if you still get the thread deadlock when importing from OSM?
- Get rid of all the changes to AbstractNodeIndexTest, that file and NodeIndex is going away with the revtree builder enhancements I'm working on.

Thanks!

emerkle826 commented:
@groldan 
- I refactored the ObjectStore concurrency tests out into their own Test class. Added a MapDB, BDB and SQLite subclass Test classes to cover them all.
- reverted all whitespace chnages
- reverted WorkingTree.java
- reverted all the Maven stress test profile stuff as it is outside the scope of this PR.

I can push the maven profile stuff to another branch if you want to pull any of it back in at a later time.
As for the OSM import tests, the best I can do is the UnmodifiableIterator approach here:
https://github.com/emerkle826/geogig/blob/dev_sqlite_storage/src/storage/sqlite/xerial/src/main/java/org/locationtech/geogig/storage/sqlite/XerialObjectDatabaseV2.java#L357-L358.

emerkle826 commented on src/storage/sqlite/xerial/src/main/java/org/locationtech/geogig/storage/sqlite/XerialObjectDatabaseV2.java:
```diff
@@ -352,10 +353,12 @@ public void putAll(Iterator<? extends RevObject> objects, final BulkOpListener l
         // System.err.println("put allllllll");
-        txHandler.startTransaction();
-        while (objects.hasNext()) {
+        //txHandler.startTransaction();
```
@groldan The change to UnmodifiableIterator seems to address the other concurrency issues I was seeing with the DataStore tests that caused me to comment out the _startTransaction()_ call here. I'm going to try to revert some of this as well and see if the tests pass now. Re-enabling this should give us better performance for when batch edits happen in large chunks.

emerkle826 replied:
Still can't wrap the SQL edits here into a single transaction. Concurrent reads still happen during a putAll that aren't able to get at the nodes _put_ in, even with **read_uncommitted** set. It's not terrible as this way, it will batch up to 50 RevObjects at a time into 1 commit transaction. So for very large numbers of RevObjects, we may have a performance penalty, but that is better than a corrupt or unpredictable database.
